File: mobile/lib/feature/map/model/nearest_station_model.dart
class NearStationModel {
  String? status;
  String? message;
  List<Data>? data;

  NearStationModel({this.status, this.message, this.data});

  NearStationModel.fromJson(Map<String, dynamic> json) {
    status = json['status'];
    message = json['message'];
    if (json['data'] != null) {
      data = <Data>[];
      json['data'].forEach((v) {
        data!.add(new Data.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['status'] = this.status;
    data['message'] = this.message;
    if (this.data != null) {
      data['data'] = this.data!.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class Data {
  String? sId;
  String? name;
  String? address;
  List<String>? images;
  String? status;
  String? availableIn;
  double? distanceKm;
  int? durationMins;
  double? rating;
  int? reviewCount;
  String? pricePerHour;
  double? pricePerHourValue;
  double? taxPercent;
  double? latitude;
  double? longitude;
  String? about;

  Data(
      {this.sId,
        this.name,
        this.address,
        this.images,
        this.status,
        this.availableIn,
        this.distanceKm,
        this.durationMins,
        this.rating,
        this.reviewCount,
        this.pricePerHour,
        this.pricePerHourValue,
        this.taxPercent,
        this.latitude,
        this.longitude,
        this.about});

  Data.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    name = json['name'];
    address = json['address'];
    images = json['images'] != null ? json['images'].cast<String>() : null;
    status = json['status'];
    availableIn = json['availableIn'];
    distanceKm = (json['distanceKm'] as num?)?.toDouble();
    durationMins = (json['durationMins'] as num?)?.toInt();
    rating = (json['rating'] as num?)?.toDouble();
    reviewCount = (json['reviewCount'] as num?)?.toInt();
    pricePerHour = json['pricePerHour']?.toString();
    pricePerHourValue = (json['pricePerHourValue'] as num?)?.toDouble();
    taxPercent = (json['taxPercent'] as num?)?.toDouble();
    latitude = (json['latitude'] as num?)?.toDouble();
    longitude = (json['longitude'] as num?)?.toDouble();
    about = json['about'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['name'] = this.name;
    data['address'] = this.address;
    data['images'] = this.images;
    data['status'] = this.status;
    data['availableIn'] = this.availableIn;
    data['distanceKm'] = this.distanceKm;
    data['durationMins'] = this.durationMins;
    data['rating'] = this.rating;
    data['reviewCount'] = this.reviewCount;
    data['pricePerHour'] = this.pricePerHour;
    data['pricePerHourValue'] = this.pricePerHourValue;
    data['taxPercent'] = this.taxPercent;
    data['latitude'] = this.latitude;
    data['longitude'] = this.longitude;
    data['about'] = this.about;
    return data;
  }
}

File: admin/UV Charging/src/models/NearestStationModel.js
import mongoose from "mongoose";

// Sub-schema: Amenity item (e.g. Restaurant, Wi-Fi, Maintenance, Shop)
const AmenitySchema = new mongoose.Schema(
  {
    label: { type: String, default: "" },  // "Restaurant", "Wi-Fi", "Maintenance", "Shop"
    icon:  { type: String, default: "" },  // icon name or image URL
  },
  { _id: false }
);

// Sub-schema: Time slot (e.g. "10:30 AM to 11:30 AM")
const SlotSchema = new mongoose.Schema(
  {
    startTime: { type: String, default: "" }, // "10:30 AM"
    endTime:   { type: String, default: "" }, // "11:30 AM"
    isBooked:  { type: Boolean, default: false },
  },
  { _id: false }
);

const NearestStationSchema = new mongoose.Schema(
  {
    // ── Basic info (used for map pins & list cards) ──
    name:        { type: String, required: true },
    address:     { type: String, default: "" },
    images:      [{ type: String }],          // carousel images array
    status: {
      type: String,
      enum: ["Available", "Unavailable", "Busy"],
      default: "Available",
    },
    availableIn: { type: String, default: "" }, // "Available in 30 minutes" (Unavailable only)
    distanceKm:  { type: Number, default: 0 },
    durationMins:{ type: Number, default: 0 },
    rating:      { type: Number, default: 0 },
    reviewCount: { type: Number, default: 0 },
    pricePerHour:{ type: String, default: "0$/hr" },  // display string e.g. "10$/hr"
    pricePerHourValue: { type: Number, default: 0 }, // numeric for calculation e.g. 10
    taxPercent:  { type: Number, default: 0 },       // tax % e.g. 5 means 5%
    latitude:    { type: Number, default: 0 },
    longitude:   { type: Number, default: 0 },

    // ── GeoJSON point for geo-queries (auto-built from latitude/longitude) ──
    location: {
      type: {
        type: String,
        enum: ["Point"],
        default: "Point",
      },
      coordinates: {
        type: [Number],   // [longitude, latitude]
        default: [0, 0],
      },
    },

    // ── Detail page fields ──
    about:       { type: String, default: "" }, // "About" section description
    amenities:   [AmenitySchema],               // Amenities list
    availableDates: [{ type: String }],         // ["06 Jan, Tue", "07 Jan, Wed", ...]
    slots:       [SlotSchema],                  // Time slots for booking

    // ── QR Code for this station ──
    qrToken:  { type: String, default: "" },   // unique token embedded in QR
    qrCode:   { type: String, default: "" },   // base64 data-URL of the QR image
  },
  {
    timestamps: true,
    versionKey: false,
  }
);

// ── 2dsphere index for geo-queries ──
NearestStationSchema.index({ location: "2dsphere" });

// ── Auto-sync location from latitude / longitude before save ──
NearestStationSchema.pre("save", function () {
  if (this.latitude && this.longitude) {
    this.location = {
      type: "Point",
      coordinates: [this.longitude, this.latitude],   // GeoJSON = [lng, lat]
    };
  }
});

// ── Also sync on findOneAndUpdate ──
NearestStationSchema.pre("findOneAndUpdate", function () {
  const update = this.getUpdate() || {};
  const lat = update?.latitude ?? update?.$set?.latitude;
  const lng = update?.longitude ?? update?.$set?.longitude;
  if (lat != null && lng != null) {
    const locationObj = { type: "Point", coordinates: [lng, lat] };
    if (update.$set) {
      update.$set.location = locationObj;
    } else {
      update.location = locationObj;
    }
    this.setUpdate(update);
  }
});

// ── Also sync on updateOne ──
NearestStationSchema.pre("updateOne", function () {
  const update = this.getUpdate() || {};
  const lat = update?.latitude ?? update?.$set?.latitude;
  const lng = update?.longitude ?? update?.$set?.longitude;
  if (lat != null && lng != null) {
    const locationObj = { type: "Point", coordinates: [lng, lat] };
    if (update.$set) {
      update.$set.location = locationObj;
    } else {
      update.location = locationObj;
    }
    this.setUpdate(update);
  }
});

const NearestStationModel = mongoose.model("NearestStation", NearestStationSchema);
export default NearestStationModel;
